File: Data Structures & Algorithms/eating-bananas/submission-0.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        left, right = 1, max(piles)
        min_speed = float("inf")

        while left <= right:
            mid = left + (right-left)//2
            total_time = self.timeTaken(piles, mid)

            if total_time > h:
                left = mid + 1
            elif total_time <= h:
                min_speed = min(min_speed, mid)
                right = mid - 1
                
        return min_speed
        

    def timeTaken(self, piles, k):
        time = 0

        for pile in piles:
            time += math.ceil(pile/k)
        
        return time

File: Data Structures & Algorithms/max-water-container/submission-0.go
func maxArea(h []int) int {
    maxArea := 0   
    left, right := 0, len(h)-1


    for left < right {
        dis := right - left
        area := min(h[left], h[right]) * dis
        maxArea = max(maxArea, area)
        if h[left] < h[right] {
            left=left+1
        } else {
            right=right-1
        }
    }
    
    return maxArea
}

func max(a, b int) int {
    if b > a {return b}
    return a
}

func min(a, b int) int {
    if b < a {return b}
    return a
}
